File: clients/ios/WSClient.swift
// WSClient.swift
// WebSocket push for clipboard.new while app is active.

import Foundation

@MainActor
final class WSClient: ObservableObject {
    @Published var isConnected = false

    var onClipboardNew: ((ClipboardEntry) -> Void)?
    var onClipboardPin: ((String, Bool) -> Void)?
    var onFilesUpdated: (() -> Void)?

    private var task: URLSessionWebSocketTask?
    private var receiveLoop = false
    private var reconnectTask: Task<Void, Never>?
    private var reconnectStep = 0
    private let reconnectSteps: [UInt64] = [1_000_000_000, 2_000_000_000, 5_000_000_000, 10_000_000_000]
    private var accessToken = ""
    private var serverURL = ""

    func connect(accessToken: String, serverURL: String) {
        self.accessToken = accessToken
        self.serverURL = serverURL
        reconnectTask?.cancel()
        reconnectTask = nil
        reconnectStep = 0
        openSocket()
    }

    func disconnect() {
        receiveLoop = false
        reconnectTask?.cancel()
        reconnectTask = nil
        reconnectStep = 0
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        isConnected = false
    }

    private func openSocket() {
        task?.cancel(with: .goingAway, reason: nil)
        task = nil

        let base = serverURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let wsBase = base.replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        guard let url = URL(string: "\(wsBase)/ws?token=\(accessToken.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? accessToken)") else {
            scheduleReconnect()
            return
        }

        task = URLSession.shared.webSocketTask(with: url)
        task?.resume()
        receiveLoop = true
        isConnected = true
        reconnectStep = 0
        listen()
        startHeartbeat()
    }

    private func listen() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .success(let message):
                if case .string(let text) = message {
                    Task { @MainActor in
                        self.handle(text)
                    }
                }
                if self.receiveLoop {
                    self.listen()
                }
            case .failure:
                Task { @MainActor in
                    self.isConnected = false
                    if self.receiveLoop {
                        self.scheduleReconnect()
                    }
                }
            }
        }
    }

    private func handle(_ text: String) {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else { return }

        if type == "pong" { return }

        if !isConnected {
            isConnected = true
            reconnectStep = 0
        }

        if type == "clipboard.new",
           let payload = json["payload"] as? [String: Any],
           let id = payload["entry_id"] as? String {
            let contentType = payload["content_type"] as? String ?? "text/plain"
            let content = payload["content"] as? String ?? ""
            let hasThumbnail = payload["has_thumbnail"] as? Bool ?? contentType.hasPrefix("image/")
            guard !content.isEmpty || hasThumbnail else { return }
            let entry = ClipboardEntry(
                id: id,
                content: content,
                contentType: contentType,
                createdAt: payload["created_at"] as? String ?? ISO8601DateFormatter().string(from: Date()),
                pinned: payload["pinned"] as? Bool ?? false,
                hasThumbnail: hasThumbnail,
                sourceDeviceId: payload["source_device_id"] as? String ?? ""
            )
            onClipboardNew?(entry)
        }

        if type == "clipboard.pin",
           let payload = json["payload"] as? [String: Any],
           let entryId = payload["entry_id"] as? String {
            let pinned = payload["pinned"] as? Bool ?? false
            onClipboardPin?(entryId, pinned)
        }

        if type == "file.ready" || type == "file.progress" {
            onFilesUpdated?()
        }
    }

    private func scheduleReconnect() {
        guard receiveLoop, reconnectTask == nil else { return }
        reconnectTask = Task { @MainActor in
            while receiveLoop && !Task.isCancelled {
                let delay = reconnectSteps[min(reconnectStep, reconnectSteps.count - 1)]
                reconnectStep = min(reconnectStep + 1, reconnectSteps.count - 1)
                try? await Task.sleep(nanoseconds: delay)
                guard receiveLoop, !Task.isCancelled else { return }
                openSocket()
                try? await Task.sleep(nanoseconds: 2_500_000_000)
                if isConnected { return }
            }
        }
    }

    private func startHeartbeat() {
        Task {
            while receiveLoop {
                try? await Task.sleep(nanoseconds: 54_000_000_000)
                task?.send(.string("{\"type\":\"ping\"}")) { _ in }
            }
        }
    }
}

struct ClipboardEntry: Identifiable, Equatable {
    let id: String
    let content: String
    let contentType: String
    let createdAt: String
    let pinned: Bool
    var hasThumbnail: Bool = false
    var sourceDeviceId: String = ""
}

File: clients/android/app/src/main/java/com/syncbridge/android/sync/WSClient.kt
package com.syncbridge.android.sync

import android.util.Log
import com.syncbridge.android.data.ApiClient
import com.syncbridge.android.data.ClipboardEntry
import com.syncbridge.android.network.NetworkManager
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject

class WSClient(
    private val api: ApiClient,
    private val networkManager: NetworkManager? = null,
    private val onClipboardNew: (ClipboardEntry) -> Unit,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var socket: WebSocket? = null
    private var reconnectJob: Job? = null
    @Volatile private var running = false
    @Volatile private var replacingSocket = false

    private val client = OkHttpClient.Builder()
        .pingInterval(30, TimeUnit.SECONDS)
        .build()

    fun connect() {
        if (running) return
        running = true
        openSocket()
    }

    private fun openSocket() {
        val token = api.accessToken ?: return
        replacingSocket = true
        socket?.close(1000, "reconnect")
        socket = null
        replacingSocket = false

        val base = api.serverUrl.trimEnd('/')
        val wsBase = base.replace("https://", "wss://").replace("http://", "ws://")
        val request = Request.Builder().url("$wsBase/ws?token=$token").build()
        socket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                reconnectJob?.cancel()
                reconnectJob = null
                setLive(true)
                Log.d(TAG, "websocket open")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val json = JSONObject(text)
                    when (json.optString("type")) {
                        "clipboard.new" -> {
                            val payload = json.getJSONObject("payload")
                            val entry = ClipboardEntry(
                                id = payload.getString("entry_id"),
                                contentType = payload.optString("content_type", "text/plain"),
                                content = payload.optString("content", ""),
                                sourceDeviceId = payload.optString("source_device_id", ""),
                                pinned = payload.optBoolean("pinned", false),
                                createdAt = payload.optString("created_at", ""),
                                hasThumbnail = payload.optBoolean("has_thumbnail", false)
                                    || payload.optString("content_type", "").startsWith("image/"),
                            )
                            onClipboardNew(entry)
                            networkManager?.markSync()
                        }
                        "signal.peer" -> {
                            val payload = json.optJSONObject("payload") ?: return
                            val deviceId = payload.optString("device_id", "")
                            if (deviceId.isBlank()) return
                            val addrsArr = payload.optJSONArray("addrs") ?: org.json.JSONArray()
                            val addrs = (0 until addrsArr.length()).map { addrsArr.getString(it) }
                            val port = payload.optInt("port", 0)
                            if (networkManager?.handleSignalPeer(deviceId, addrs, port) == true) {
                                SyncEventBus.emitNearbyAlert(deviceId)
                            }
                        }
                        "file.ready", "file.progress" -> {
                            networkManager?.markSync()
                            if (json.optString("type") == "file.ready") {
                                SyncEventBus.emitFilesUpdated()
                            }
                        }
                    }
                } catch (e: Exception) {
                    Log.w(TAG, "ws parse: ${e.message}")
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                if (replacingSocket) return
                Log.d(TAG, "websocket closed code=$code reason=$reason")
                handleDisconnect()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (replacingSocket) return
                Log.w(TAG, "websocket failure: ${t.message}")
                handleDisconnect()
            }
        })
    }

    private fun handleDisconnect() {
        setLive(false)
        scheduleReconnect()
    }

    private fun setLive(live: Boolean) {
        SyncEventBus.setConnected(live)
        networkManager?.setWsConnected(live)
    }

    private fun scheduleReconnect() {
        if (!running || reconnectJob?.isActive == true) return
        reconnectJob = scope.launch {
            var backoff = 1000L
            while (running && !SyncEventBus.connected.value) {
                delay(backoff)
                if (!running) return@launch
                openSocket()
                delay(2500)
                if (SyncEventBus.connected.value) return@launch
                backoff = minOf(backoff * 2, 30_000L)
            }
        }
    }

    fun disconnect() {
        running = false
        reconnectJob?.cancel()
        reconnectJob = null
        replacingSocket = true
        socket?.close(1000, "stop")
        socket = null
        replacingSocket = false
        setLive(false)
    }

    companion object {
        private const val TAG = "SyncBridgeWS"
    }
}
